bigchaindb/core.py

```diff
@@ -553,41 +556,54 @@ def vote(self, block, previous_block_id, decision, invalid_reason=None):
 
         return vote_signed
 
-    def write_vote(self, block, vote, block_number):
+    def write_vote(self, block, vote):
         """Write the vote to the database."""
 
         # First, make sure this block doesn't contain a vote from this node
         if self.has_previous_vote(block):
             return None
 
-        # We need to *not* override the existing block_number, if any
-        # FIXME: MIGHT HAVE RACE CONDITIONS WITH THE OTHER NODES IN THE FEDERATION
-        if 'block_number' not in vote:
-            vote['block_number'] = block_number  # maybe this should be in the signed part...or better yet, removed..
-
         r.table('votes') \
             .insert(vote) \
             .run(self.conn)
 
     def get_last_voted_block(self):
         """Returns the last block that this node voted on."""
 
-        last_voted = r.table('votes') \
-            .filter(r.row['node_pubkey'] == self.me) \
-            .order_by(r.desc('block_number')) \
-            .limit(1) \
-            .run(self.conn)
-
-        # return last vote if last vote exists else return Genesis block
-        if not last_voted:
+        try:
+            # get the latest value for the vote timestamp (over all votes)
+            max_timestamp = r.table('votes') \
+                .filter(r.row['node_pubkey'] == self.me) \
+                .concat_map(lambda x: [x['vote']['timestamp']]) \
+                .max() \
+                .run(self.conn)
+
+            last_voted = list(r.table('votes') \
+                .filter(r.row['node_pubkey'] == self.me) \
+                .filter(lambda x: x['vote']['timestamp'] == max_timestamp) \
+                .run(self.conn))
+
+        except r.ReqlNonExistenceError:
+            # return last vote if last vote exists else return Genesis block
             return list(r.table('bigchain')
-                        .filter(r.row['block_number'] == 0)
+                        .filter(util.is_genesis_block)
                         .run(self.conn))[0]
 
-        res = r.table('bigchain').get(last_voted[0]['vote']['voting_for_block']).run(self.conn)
+        # since the resolution of timestamp is a second,
+        # we might have more than one vote per timestamp
+        if len(last_voted) > 1:
```

Comment:
the inner algorithm seems to work fine if the length `last_voted` is 1 so we may just remove the `if else` statement